File: JatxMusicFX/MusicCommons/src/main/java/jatx/musiccommons/transmitter/threads/TransmitterController.java
/*******************************************************************************
 * Copyright (c) 2015 Evgeny Tabatsky.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 * 
 * Contributors:
 *     Evgeny Tabatsky - initial API and implementation
 ******************************************************************************/
package jatx.musiccommons.transmitter.threads;

import jatx.musiccommons.transmitter.Globals;
import jatx.musiccommons.transmitter.UI;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.concurrent.ConcurrentHashMap;

public class TransmitterController extends Thread {
	public static final int CONNECT_PORT_CONTROLLER = 7172;
	
	volatile WeakReference<UI> uiRef;

	private volatile int volume = 0;

	private volatile ConcurrentHashMap<String, TransmitterControllerWorker> workers =
			new ConcurrentHashMap<>();

	volatile boolean finishFlag;
	
	volatile boolean mForceDisconnectFlag;
	
	public TransmitterController(UI ui) {
		finishFlag = false;
		mForceDisconnectFlag = false;
		
		uiRef = new WeakReference<>(ui);
	}
	
	public void setFinishFlag() {
		finishFlag = true;
	}
	
	public void forceDisconnect() {
		mForceDisconnectFlag = true;
	}

	public void setVolume(int volume) {
		final String msg = "(controller) set volume: " + volume;
		System.out.println(msg);
		workers.forEachValue(0L, transmitterControllerWorker -> transmitterControllerWorker.setVolume(volume));
		this.volume = volume;
	}

	public void play() {
		System.out.println("(controller) play");
		workers.forEachValue(0L, transmitterControllerWorker -> transmitterControllerWorker.play());
	}

	public void pause() {
		System.out.println("(controller) pause");
		workers.forEachValue(0L, transmitterControllerWorker -> transmitterControllerWorker.pause());
	}
	
	@Override
	public void run() {
		ServerSocket ss = null;
		
		try {
			while(!finishFlag) {
				Thread.sleep(100);

				ss = new ServerSocket(CONNECT_PORT_CONTROLLER);
				System.out.println("(controller) new server socket");

				try {
					ss.setSoTimeout(Globals.SO_TIMEOUT);
					Socket s = ss.accept();
					System.out.println("(controller) server socket accept");
					String host = s.getInetAddress().getHostAddress();
					TransmitterControllerWorker worker = new TransmitterControllerWorker(s, host);
					worker.start();
					System.out.println("(controller) worker $it started");
					workers.put(host, worker);
					final String msg = "(controller) total workers: " + workers.size();
					System.out.println(msg);
					worker.onWorkerStopped = () -> {
						System.out.println("(controller) worker $host stopped");
						workers.remove(host);
						final String msg2 = "(player) total workers: " + workers.size();
						System.out.println(msg2);
					};
					setVolume(volume);
				} catch (SocketTimeoutException e) {
					System.out.println("(controller) socket timeout");
				} finally {
					try {
						ss.close();
						System.out.println("(controller) server socket closed");
					} catch (IOException e) {
						e.printStackTrace();
					}
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		} catch (InterruptedException e) {
			System.err.println("(controller) thread interrupted");
		} finally {
			workers.forEachValue(0L, transmitterControllerWorker -> transmitterControllerWorker.setFinishWorkerFlag());
			System.out.println("(controller) workers finished");
			try {
				ss.close();
				System.out.println("(controller) server socket closed");
			} catch (Exception ex) {
				System.err.println("(controller) cannot close server socket");
			}
			System.out.println("(controller) thread finished");
		}
	}
}

File: JatxMusicAndroid2/musicTransmitterAndroid/src/main/java/jatx/musictransmitter/android/threads/TransmitterController.kt
package jatx.musictransmitter.android.threads

import jatx.debug.logError
import jatx.musictransmitter.android.data.Settings
import java.io.IOException
import java.io.OutputStream
import java.net.ServerSocket
import java.net.SocketTimeoutException
import java.util.*
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue

const val CONNECT_PORT_CONTROLLER = 7172

const val COMMAND_EMPTY = 255.toByte()
const val COMMAND_STOP = 127.toByte()
const val COMMAND_PAUSE = 126.toByte()
const val COMMAND_PLAY = 125.toByte()

const val SO_TIMEOUT = 1000

class TransmitterController(
    private val initialVolume: Int
) : Thread() {
    @Volatile
    var finishFlag = false
    @Volatile
    var fifo: BlockingQueue<Byte> = ArrayBlockingQueue<Byte>(2048)

    @Volatile
    var forceDisconnectFlag = false

    var tp: TransmitterPlayer? = null

    private var ss: ServerSocket? = null
    private var os: OutputStream? = null

    fun play() {
        println("(controller) play")
        fifo.offer(COMMAND_PLAY)
    }

    fun pause() {
        println("(controller) pause")
        fifo.offer(COMMAND_PAUSE)
    }

    fun sendStop() {
        println("(controller) stop")
        fifo.offer(COMMAND_STOP)
    }

    fun setVolume(vol: Int) {
        println("(controller) set volume: " + Integer.valueOf(vol).toString())
        if (vol in 0..100) {
            fifo.offer(vol.toByte())
        }
    }

    override fun run() {
        println("(controller) thread started")
        try {
            while (!finishFlag) {
                sleep(100)
                ss = ServerSocket(CONNECT_PORT_CONTROLLER)
                println("(controller) new server socket")
                try {
                    ss?.soTimeout = SO_TIMEOUT
                    val s = ss?.accept()
                    os = s?.getOutputStream()
                    println("(controller) socket connect")
                    setVolume(initialVolume)
                    while (!finishFlag) {
                        val cmd = fifo.poll() ?: COMMAND_EMPTY
                        val data = byteArrayOf(cmd)
                        os?.write(data)
                        os?.flush()
                        sleep(10)
                        if (forceDisconnectFlag) {
                            println("(player) disconnect flag: throwing DisconnectException")
                            throw ForceDisconnectException()
                        }
                    }
                    val data = byteArrayOf(COMMAND_STOP)
                    os?.write(data)
                    os?.flush()
                } catch (e: SocketTimeoutException) {
                    println("(controller) socket timeout")
                } catch (e: ForceDisconnectException) {
                    println("(controller) socket force disconnect")
                    println("(controller) " + Date().time % 10000)
                } catch (e: IOException) {
                    println("(controller) socket disconnect")
                    println("(controller) " + Date().time % 10000)
                    tp?.forceDisconnectFlag = true
                    sleep(250)
                } finally {
                    forceDisconnectFlag = false
                    os?.close()
                    println("(controller) outstream closed")
                    ss?.close()
                    println("(controller) server socket closed")
                }
            }
        } catch (e: IOException) {
            logError(e)
        } catch (e: InterruptedException) {
            println("(controller) thread interrupted")
            os?.close()
            println("(controller) outstream closed")
            ss?.close()
            println("(controller) server socket closed")
        } finally {
            println("(controller) thread finished")
        }
    }
}